### PR TITLE
immutable aggregate state

### PR DESCRIPTION
**Pull request by andreasBerre**
- fixed naming
- supporting immutable aggregate states, and that commands and events can be used as either applicators or initializers depending on the state of the aggregate

### CODE REVIEW

**Comment by mortenaa** [comment]
looks good! would it make sense for sagas to allow the same events used as applicators or initializers depending on saga state also?

**Comment by andreasBerre** [comment]
@mortenaa 
> looks good! would it make sense for sagas to allow the same events used as applicators or initializers depending on saga state also?

good idea. Will take a look, will implement if its not too much trouble.

File: k-es-lib/src/test/kotlin/no/ks/kes/lib/SagaConfigurationTest.kt [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+
+        "test that creating a saga with multiple initializers throws exception" {
+            shouldThrow<IllegalStateException> {
+                object : Saga<SomeState>(SomeState::class, "SomeSaga") {
+                    init {
+                        init<SomeEvent>({ it.aggregateId }) { SomeState(it.aggregateId) }
+                        init<SomeEvent>({ it.aggregateId }) { SomeState(it.aggregateId) }
+                    }
+                }
+            }.apply {
```
Comment (zapodot) [inline k-es-lib/src/test/kotlin/no/ks/kes/lib/SagaConfigurationTest.kt:31]: May be simplified. Why not do something like:
```kotlin
shouldThrow<IllegalStateException> {
...
}.message shouldContain "Duplicate init handler" 
```

The use of an apply function sort of implies that you are making some change to the state of the object which you do not here (also used in other tests)

**Comment by zapodot** [comment]
As suggested in a comment on one of your tests: drop the use of .apply for shouldThrow<> expressions as it looks a bit confusing. Otherwise the change looks good!

File: k-es-lib/src/test/kotlin/no/ks/kes/lib/SagaConfigurationTest.kt (hunk 1 above)
Comment (andreasBerre) [reply k-es-lib/src/test/kotlin/no/ks/kes/lib/SagaConfigurationTest.kt:31]: agreed, scoping func not necessary as long as only one assert is executed, and even then apply is probably not the best choice: "run" is better maybe?

Comment (zapodot) [reply k-es-lib/src/test/kotlin/no/ks/kes/lib/SagaConfigurationTest.kt:31]: Yes "run" would be better if you have more than one assertion

**Pull request merged**